system/HTTP/SiteURIFactory.php

```diff
@@ -0,0 +1,224 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\HTTP;
+
+use Config\App;
+
+class SiteURIFactory
+{
+    /**
+     * @var array Superglobal SERVER array
+     */
+    private array $server;
+
+    /**
+     * @var array Superglobal GET array
+     */
+    private array $get;
+
+    private App $appConfig;
+
+    /**
+     * @param array $server Superglobal $_SERVER array
+     * @param array $get    Superglobal $_GET array
+     */
+    public function __construct(array &$server, array &$get, App $appConfig)
+    {
+        $this->server    = &$server;
+        $this->get       = &$get;
```

Comment:
I'm using references. but since we have not used pass-by-reference for superglobals before, 
this may be breaking changes. I'll revert back to passing by values.

Reply:
Done.